### PR TITLE
Add categorical distribution

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #4678.

### CODE REVIEW

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: tests/chainer_tests/distributions_tests/test_kldivergence.py [hunk 1]
```diff
@@ -63,6 +70,18 @@ def make_normal_dist(self, is_gpu=False, use_log_scale=False):
             params = self.encode_params({"loc": loc, "scale": scale}, is_gpu)
         return distributions.Normal(**params)
 
+    @testing.with_requires('numpy>=1.11')
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_kldivergence.py]: Could you explain why new numpy is required?

Comment (YoshikawaMasashi) [reply tests/chainer_tests/distributions_tests/test_kldivergence.py]: In old numpy, this error occurs,
```
>>> import numpy
>>> p = numpy.random.normal(size=(3)).astype(numpy.float32)
>>> p = numpy.exp(p)
>>> p /= p.sum()
>>> numpy.random.choice(3, p=p)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "mtrand.pyx", line 1083, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:10106)
ValueError: probabilities do not sum to 1
```

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: chainer/distributions/__init__.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 """Collection of distribution implementations."""
 
+from chainer.distributions.categorical import Categorical  # NOQA
```
Comment (toslunar) [inline chainer/distributions/__init__.py]: Could you make import sorted?

File: docs/source/reference/distributions.rst [hunk 3]
```diff
@@ -14,6 +14,7 @@ Distributions
    :toctree: generated/
    :nosignatures:
    
+   chainer.distributions.Categorical
```
Comment (toslunar) [inline docs/source/reference/distributions.rst]: Ditto. (sort)

**Comment by toslunar** [comment]
Could you suppress warnings (`RuntimeWarning: divide by zero encountered in log`)?  If it is intended, please fix it like #5122.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: chainer/distributions/categorical.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    def batch_shape(self):
+        return self.p.shape[:-1]
+
+    @property
+    def event_shape(self):
+        return ()
+
+    @property
+    def entropy(self):
+        print(self.p)
```
Comment (toslunar) [inline chainer/distributions/categorical.py]: Could you remove the line?

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**